rdflib/plugins/serializers/hext.py

```diff
@@ -34,70 +35,85 @@ def __init__(self, store: Union[Graph, ConjunctiveGraph]):
-        _iri_or_bn(context) if not str(context).startswith(("_", "file://")) else '""'
-    )
-
+                hl = self._hex_line(triple, context)
+                if hl is not None:
+                    stream.write(hl.encode())
 
-def _iri_or_bn(i_):
-    if type(i_) == URIRef:
-        return f"\"{i_}\""
-    else:
-        return f"\"{i_.n3()}\""
+    def _hex_line(self, triple, context):
+        if type(triple[0]) in [URIRef, BNode]:  # exclude QuotedGraph and other objects
+            # value
+            value = triple[2] \
+                if type(triple[2]) == Literal \
+                else self._iri_or_bn(triple[2])
 
+            # datatype
+            if type(triple[2]) == URIRef:
+                # datatype = "http://www.w3.org/1999/02/22-rdf-syntax-ns#namedNode"
+                datatype = "globalId"
+            elif type(triple[2]) == BNode:
+                # datatype = "http://www.w3.org/1999/02/22-rdf-syntax-ns#blankNode"
+                datatype = "localId"
+            elif type(triple[2]) == Literal:
+                if triple[2].datatype is not None:
+                    datatype = f"{triple[2].datatype}"
+                else:
+                    if triple[2].language is not None:  # language
+                        datatype = RDF.langString
+                    else:
+                        datatype = XSD.string
+            else:
+                return None  # can't handle non URI, BN or Literal Object (QuotedGraph)
 
-def _literal(i_):
-    raw_datatype = [
-        "http://www.w3.org/2001/XMLSchema#integer",
-        "http://www.w3.org/2001/XMLSchema#long",
-        "http://www.w3.org/2001/XMLSchema#int",
-        "http://www.w3.org/2001/XMLSchema#short",
-        "http://www.w3.org/2001/XMLSchema#positiveInteger",
-        "http://www.w3.org/2001/XMLSchema#negativeInteger",
-        "http://www.w3.org/2001/XMLSchema#nonPositiveInteger",
-        "http://www.w3.org/2001/XMLSchema#nonNegativeInteger",
-        "http://www.w3.org/2001/XMLSchema#unsignedLong",
-        "http://www.w3.org/2001/XMLSchema#unsignedInt",
-        "http://www.w3.org/2001/XMLSchema#unsignedShort",
+            # language
+            if type(triple[2]) == Literal:
+                if triple[2].language is not None:
+                    language = f"{triple[2].language}"
+                else:
+                    language = ""
+            else:
+                language = ""
 
-        "http://www.w3.org/2001/XMLSchema#float",
-        "http://www.w3.org/2001/XMLSchema#double",
-        "http://www.w3.org/2001/XMLSchema#decimal",
+            return '["%s", "%s", "%s", "%s", "%s", "%s"]\n' % (
+                self._iri_or_bn(triple[0]),
+                triple[1],
+                value,
+                datatype,
+                language,
+                self._context(context)
+            )
+        else:  # do not return anything for non-IRIs or BNs, e.g. QuotedGraph, Subjects
+            return None
```

Comment:
Can the type comparisons be done using `isinstance()`? Or is there a requirement to use `type()`?

Reply:
Is there an advantage to using `isinstance()`? No real reason not to: that's just what I started doing! Happy to change if `isinstance()` is _better_ in some way.

Reply:
I presume you're concerned with type inheritance, in case people use subclasses of, say, `Literal`? I'll convert (and solve the reamining failing test).

Reply:
`isinstance()` caters for subclasses while `type()` does not. 

```python
class SpecialURIRef(URIRef):
    pass


special_uri = SpecialURIRef("123")

type(special_uri) == URIRef
>>> False

isinstance(special_uri, URIRef)
>>> True
```

Using `type()` could lead to similar issues like https://github.com/RDFLib/rdflib/issues/1332

Reply:
Done in Commit https://github.com/RDFLib/rdflib/pull/1489/commits/75b797e0aacdd098d5ad7c863fc628d468f4dfe9